src/theory/arith/nonlinear_extension.cpp

```diff
@@ -1008,126 +796,19 @@ int NonlinearExtension::checkLastCall(const std::vector<Node>& assertions,
-          }
-        }
-        else
-        {
-          // new representative of congruence class
-          d_funcMap[ak].push_back(a);
-        }
-        // add to congruence class
-        d_funcCongClass[aa].push_back(a);
-      }
-    }
-    else if (ak == PI)
-    {
-      Assert(consider);
-      needPi = true;
-      d_funcMap[ak].push_back(a);
-      d_funcCongClass[a].push_back(a);
-    }
-    else
-    {
-      Assert(false);
-    }
-  }
-  // initialize pi if necessary
-  if (needPi && d_pi.isNull())
-  {
-    mkPi();
-    getCurrentPiBounds(lemmas);
   }
 
+  // initialize the trancendental function solver
+  std::vector<Node> lemmas;
+  d_trSlv.initLastCall(assertions, false_asserts, xts, lemmas, lemsPp);
+
+  // process lemmas that may have been generated by the transcendental solver
   filterLemmas(lemmas, lems);
-  if (!lems.empty())
+  if (!lems.empty() || !lemsPp.empty())
   {
     Trace("nl-ext") << "  ...finished with " << lems.size()
                     << " new lemmas during registration." << std::endl;
-    return lems.size();
-  }
-
-  // process SINE phase shifting
-  for (const Node& a : trNeedsMaster)
-  {
-    // should not have processed this already
-    Assert(d_trMaster.find(a) == d_trMaster.end());
-    Kind k = a.getKind();
-    Assert(k == SINE || k == EXPONENTIAL);
-    Node y =
-        nm->mkSkolem("y", nm->realType(), "phase shifted trigonometric arg");
-    Node new_a = nm->mkNode(k, y);
-    d_trSlaves[new_a].insert(new_a);
-    d_trSlaves[new_a].insert(a);
-    d_trMaster[a] = new_a;
-    d_trMaster[new_a] = new_a;
-    Node lem;
-    if (k == SINE)
-    {
-      Trace("nl-ext-tf") << "Basis sine : " << new_a << " for " << a
-                         << std::endl;
-      Assert(!d_pi.isNull());
-      Node shift = nm->mkSkolem("s", nm->integerType(), "number of shifts");
-      // TODO : do not introduce shift here, instead needs model-based
-      // refinement for constant shifts (cvc4-projects #1284)
-      lem = nm->mkNode(
-          AND,
-          mkValidPhase(y, d_pi),
-          nm->mkNode(
-              ITE,
-              mkValidPhase(a[0], d_pi),
-              a[0].eqNode(y),
-              a[0].eqNode(nm->mkNode(
-                  PLUS,
-                  y,
-                  nm->mkNode(MULT, nm->mkConst(Rational(2)), shift, d_pi)))),
-          new_a.eqNode(a));
-    }
-    else
-    {
-      // do both equalities to ensure that new_a becomes a preregistered term
-      lem = nm->mkNode(AND, a.eqNode(new_a), a[0].eqNode(y));
-    }
-    // note we must do preprocess on this lemma
-    Trace("nl-ext-lemma") << "NonlinearExtension::Lemma : purify : " << lem
-                          << std::endl;
-    lemsPp.push_back(lem);
-  }
-  if (!lemsPp.empty())
-  {
-    Trace("nl-ext") << "  ...finished with " << lemsPp.size()
-                    << " new lemmas SINE phase shifting." << std::endl;
-    return lemsPp.size();
+    return lems.size() + lemsPp.size();
```

Comment:
why are we summing them up? aren't lems and lemsPp sematically the same lemmas -- the only difference is the preprocessing, right?

Reply:
`lems` and `lemsPp` are two independent sets of lemmas.

The code was modified slightly here. Previously, transcendental function handling in this block of code generated lems and then sent out lemmas if any were generated. If no lemmas were generated, it proceeded to generate a new set of lemmas in lemsPp.  Since the transcendental function solver does not have access to the output channel, these two steps are combined into one return case here.

Reply:
ok, got it